File: packages/sdkwork-utils-php/src/EncodingUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class EncodingUtils
{
    public static function base64Encode(string $value): string
    {
        return base64_encode($value);
    }

    public static function base64Decode(string $value): ?string
    {
        $decoded = base64_decode(trim($value), true);
        return $decoded === false ? null : $decoded;
    }

    public static function hexEncode(string $value): string
    {
        return bin2hex($value);
    }

    public static function hexDecode(string $value): ?string
    {
        $trimmed = trim($value);
        if ($trimmed === '' || strlen($trimmed) % 2 !== 0) {
            return null;
        }
        $decoded = hex2bin($trimmed);
        return $decoded === false ? null : $decoded;
    }

    public static function urlEncode(string $value): string
    {
        return rawurlencode($value);
    }

    public static function urlDecode(string $value): ?string
    {
        return rawurldecode($value);
    }

    public static function base64urlEncode(string $value): string
    {
        return rtrim(strtr(base64_encode($value), '+/', '-_'), '=');
    }

    public static function base64urlDecode(string $value): ?string
    {
        $trimmed = trim($value);
        if ($trimmed === '') {
            return null;
        }
        $padded = str_pad($trimmed, strlen($trimmed) + (4 - strlen($trimmed) % 4) % 4, '=', STR_PAD_RIGHT);
        $decoded = base64_decode(strtr($padded, '-_', '+/'), true);
        return $decoded === false ? null : $decoded;
    }
}

File: packages/sdkwork-utils-csharp/EncodingUtils.cs
namespace Sdkwork.Utils;

public static class EncodingUtils
{
    public static string Base64Encode(byte[] value) => Convert.ToBase64String(value);
    public static string Base64Encode(string value) => Base64Encode(System.Text.Encoding.UTF8.GetBytes(value));

    public static byte[]? Base64Decode(string value)
    {
        try
        {
            return Convert.FromBase64String(value.Trim());
        }
        catch (FormatException)
        {
            return null;
        }
    }

    public static string HexEncode(byte[] value) => Convert.ToHexString(value).ToLowerInvariant();

    public static byte[]? HexDecode(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length % 2 != 0) return null;
        try
        {
            return Convert.FromHexString(trimmed);
        }
        catch (FormatException)
        {
            return null;
        }
    }

    public static string UrlEncode(string value) => Uri.EscapeDataString(value);

    public static string? UrlDecode(string value)
    {
        try
        {
            return Uri.UnescapeDataString(value);
        }
        catch (UriFormatException)
        {
            return null;
        }
    }

    public static string Base64UrlEncode(byte[] value) =>
        Convert.ToBase64String(value).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static string Base64UrlEncode(string value) => Base64UrlEncode(System.Text.Encoding.UTF8.GetBytes(value));

    public static byte[]? Base64UrlDecode(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }

        var padded = trimmed.PadRight(trimmed.Length + (4 - trimmed.Length % 4) % 4, '=')
            .Replace('-', '+')
            .Replace('_', '/');
        try
        {
            return Convert.FromBase64String(padded);
        }
        catch (FormatException)
        {
            return null;
        }
    }
}
